sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/NumberUtils.scala

```diff
@@ -0,0 +1,189 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.util
+
+import java.math.BigDecimal
+import java.text.{DecimalFormat, NumberFormat, ParsePosition}
+import java.util.Locale
+
+import org.apache.spark.sql.errors.{QueryCompilationErrors, QueryExecutionErrors}
+import org.apache.spark.sql.types.Decimal
+import org.apache.spark.unsafe.types.UTF8String
+
+object NumberUtils {
+
+  private val pointSign = '.'
+  private val letterPointSign = 'D'
+  private val commaSign = ','
+  private val letterCommaSign = 'G'
+  private val minusSign = '-'
+  private val letterMinusSign = 'S'
+  private val dollarSign = '$'
+
+  private val commaSignStr = commaSign.toString
+
+  private def normalize(format: String): String = {
+    var notFindDecimalPoint = true
+    val normalizedFormat = format.toUpperCase(Locale.ROOT).map {
+      case '9' if notFindDecimalPoint => '#'
+      case '9' if !notFindDecimalPoint => '0'
+      case `letterPointSign` =>
+        notFindDecimalPoint = false
+        pointSign
+      case `letterCommaSign` => commaSign
+      case `letterMinusSign` => minusSign
+      case `pointSign` =>
+        notFindDecimalPoint = false
+        pointSign
+      case other => other
+    }
+    // If the comma is at the beginning or end of number format, then DecimalFormat will be invalid.
+    // For example, "##,###," or ",###,###" for DecimalFormat is invalid, so we must use "##,###"
```

Comment:
This is a bit confusing. If it's invalid, shall we fail instead of silently removing the starting/ending comma?

Reply:
For example: PostgreSQL and Oracle support `select to_number('12,454,', '99,999,')`;
'99,999,' will be normalized as "##,###,". I remove the last ',' , so this PR could supports it too.